### PR TITLE
 [#160718635] auth user signup

### PR DESCRIPTION
**Pull request by akpante3**
#### What does this PR do
it contains an API endpoint for signing up a user

#### Description of Task to be completed?
The page contains the following
1. setup database
1. created an endpoint where users can sign up 
2. wrote   test for the sign up endpoint
3. tested Endpoint on postman

#### How should this be manually tested?
on console
. cd server
. npm start

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
(https://www.pivotaltracker.com/story/show/160718635)

#### Screenshot
N/A

success in sign up

#### Questions:
N/A

### CODE REVIEW

File: server/app/config/index.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const env = process.env.NODE_ENV || 'development';
+console.log(process.env.PORT);
```
Comment (dogoyaro) [inline server/app/config/index.js]: Please remove console.logs

File: server/app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * @param {string} email
+ * @param {string} password
+ * @param {string} user
+ * @param {string} address
+ * @return {string} token
+ * @public
+*/
+const createUser = (email, password, user, address) => {
+  const hashedPassword = bcrypt.hashSync(password, 8);
+  return db.one(`INSERT INTO users (email, password, name, address)
```
Comment (dogoyaro) [inline server/app/controllers/users.js:16]: Wouldn't it be better to abstract your database calls away into db helper functions?

File: server/app/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          expiresIn: 86400,
+        }
+      );
+      return Promise.resolve({
+        id: data.id,
+        email: data.email,
+        name: data.name,
+        address: data.address,
+        token,
+
```
Comment (dogoyaro) [inline server/app/controllers/users.js]: space is unnecessary

File: server/app/controllers/users.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/**  create a User
+ * @param {string} email
+ * @param {string} password
+ * @param {string} user
+ * @param {string} address
+ * @return {string} token
+ * @public
+*/
+const createUser = (email, password, user, address) => {
+  const hashedPassword = bcrypt.hashSync(password, 8);
```
Comment (dogoyaro) [inline server/app/controllers/users.js:15]: You are not calling this function appropriately according to the [documentation](https://www.pivotaltracker.com/n/projects/2196560).

File: server/app/middlewares/auth.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import validator from 'email-validator';
```
Comment (dogoyaro) [inline server/app/middlewares/auth.js]: You cannot use this validator. You have to write your own custom email validator.

File: server/app/middlewares/auth.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+
+const authNewUser = (req, res, next) => {
+  const username = req.body.username.trim();
+
+  if (!req.body.email || !req.body.password || !req.body.username || !req.body.address) {
```
Comment (dogoyaro) [inline server/app/middlewares/auth.js]: The exact field that is missing from the input might be more helpful here for the user. What do you think?

File: server/app/middlewares/auth.js (hunk 5 above)
Comment (dogoyaro) [inline server/app/middlewares/auth.js]: I think the name of this file might be a little misleading as it doesn't actually authenticate but only validates user input

File: server/app/middlewares/auth.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    return res.status(400).send({
+      status: 'failure',
+      message: 'invalid email or username,please input a valid email',
+    });
+  }
+
+  next();
+};
+
+export default authNewUser;
```
Comment (dogoyaro) [inline server/app/middlewares/auth.js]: Remember that the default export of a file should match the name of the file. Please review the [airbnb-styleguide](https://github.com/airbnb/javascript)

File: server/app/middlewares/users.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+      });
+    }).catch((error) => {
+      res.status(400).send({
+        status: 'failure',
+        message: error.message,
+      });
+    });
+};
+
+export default newUsers;
```
Comment (dogoyaro) [inline server/app/middlewares/users.js:20]: The default export should match the name of the file.

File: server/app/middlewares/users.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import createUser from './../controllers/users';
```
Comment (dogoyaro) [inline server/app/middlewares/users.js:1]: I think this logic should be done in the controller and not in a middle-ware. Please share the reason why you have decided to do it this way

File: server/app/tests/test.test.js [hunk 10]
```diff
@@ -1,189 +1,62 @@
...
-      .get('/api/v1/orders/65')
-      .set('Accept', 'application/json')
-      .expect(200)
-      .expect((res) => {
-        expect(res.body.status === 'success');
-      })
-      .end((err) => {
-        if (err) return done(err);
-        return done();
+      .expect()
```
Comment (dogoyaro) [inline server/app/tests/test.test.js]: You are making no assertions in this test. Please fix.

File: server/app/tests/test.test.js [hunk 11]
```diff
@@ -1,189 +1,62 @@
...
+      .send(user)
       .expect(400)
-      .expect((res) => {
-        expect(res.body.status === 'failure');
-      })
-      .end((err) => {
-        if (err) return done(err);
-        return done();
-      });
+      .end(done);
```
Comment (dogoyaro) [inline server/app/tests/test.test.js:75]: You're also making no assertions in this test case. Please fix.

**Review by dogoyaro — CHANGES_REQUESTED** [review]

**Comment by dogoyaro** [comment]
Please use proper markdown as described in the slack channel for the relevant pivotal tracker stories.

File: server/app/tests/test.test.js (hunk 11 above)
Comment (akpante3) [reply server/app/tests/test.test.js:75]: fixed

File: server/app/tests/test.test.js (hunk 10 above)
Comment (akpante3) [reply server/app/tests/test.test.js]: fixed

File: server/app/middlewares/users.js (hunk 9 above)
Comment (akpante3) [reply server/app/middlewares/users.js:1]: I learned it from my LFA in the last boot camp, please what better way can it be done?

File: server/app/middlewares/auth.js (hunk 7 above)
Comment (akpante3) [reply server/app/middlewares/auth.js]: fixed

File: server/app/middlewares/auth.js (hunk 5 above)
Comment (akpante3) [reply server/app/middlewares/auth.js]: done

File: server/app/controllers/users.js (hunk 4 above)
Comment (akpante3) [reply server/app/controllers/users.js:15]: the documentation you sent is for pivotal tracker

File: server/app/controllers/users.js (hunk 3 above)
Comment (akpante3) [reply server/app/controllers/users.js]: fixed

File: server/app/config/index.js (hunk 1 above)
Comment (akpante3) [reply server/app/config/index.js]: fixed

**Pull request merged**